### ISSUE TITLE
Auth for sites with authentication

### ISSUE THREAD

**Issue by vromero** [dkhamsing/awesome_bot#105]
Support for authentication, at least basic would be great.

Support for github auth would be great too.

**Comment by dkhamsing** [comment]
Can you elaborate

Do you mean a private repo on GitHub?

**Comment by vromero** [comment]
It would be great to support two auth approaches, 

one would be to provide a github login/pass for any github link (that would probably be the most common use-case), perhaps bitbucket too?

another would be the ability to prepend user:password to some given linkgs for instance --basic-auth https://username:password@somedomain.com (will add basic auth with username:password to any link inside somedomain.com, the param name is just brainstorming).

**Comment by dkhamsing** [comment]
afaik the majority use has been in public projects on github so you can see how this would not work as the credentials would be exposed

but if you are using this privately on a machine you control, it could work?

would gladly accept any pull requests :)

**Comment by vromero** [comment]
I wish I were proficient enough in Ruby to contribute. Plus on a second though I would had it implemented with curl cookies, so people can log in even in custom websites with curl and then reuse the cookies.

**Comment by woneill** [comment]
I needed the same for validating links in a private github repo. I submitted #134 to handle the basic auth idea from https://github.com/dkhamsing/awesome_bot/issues/105#issuecomment-249938818.

I've had to use raw.githubusercontent.com as the base_url tho like so:

```
awesome_bot --base-url https://woneill:PersonalAccessToken@raw.githubusercontent.com/organization/private-repo/master
```

**Comment by dkhamsing** [comment]
@woneill so to clarify, you can only check one private site per readme?

**Comment by woneill** [comment]
I updated the URL parsing logic to use the username and password in URLs like https://username:password@somedomain.com. I think the only way to securely use that w/o having embedded credentials would be via `--base_url`

**Comment by dkhamsing** [comment]
Ah I see

So this would work for what @vromero proposes but your suggestion is to not have this in the readme, but rather pass it as a command argument

**Comment by woneill** [comment]
Sorry- I might have been over-explaining. I didn't add any additional command arguments- I just updated the URL parser logic to extract the username:password when part of a URL and using them as basic auth credentials.

**Comment by dkhamsing** [comment]
yes yes, it is what I meant also

**Comment by vromero** [comment]
> Plus on a second though I would had it implemented with curl cookies, so people can log in even in custom websites with curl and then reuse the cookies.
Actually I changed my mind later on.
I *believe* it would be more reasonable to just let the user manually log in to as many domains as he wants (saving into a cookie file) and then pass the cookie file to awesome_bot.

**Comment by dkhamsing** [comment]
hm interesting

**Comment by dkhamsing** [comment]
@vromero please open another issue if still interested in the cookies idea, thx

**Comment by dkhamsing** [comment]
this was released with `1.17.0` published today

**Comment by vromero** [comment]
I still feels the cookie is the most favorable solution as for a reasonable common use case:

- A number of a.md, b.md, c.md files in github
- Links a->b, b->c, c->a

This solution would only help to get to the mark down desired (a or b or c) but will not help with the links (unless I'm wrong). To have login and password in the markdowns I don't think is a solution for anybody.

I'll open another issue for the cookies.
